### DISCUSSION TITLE
Undefined reference issue, can't link right the project, cmake + mingw

### DISCUSSION THREAD

**Discussion by lazzybunny** [ocornut/imgui#7845] [New Users Build/Link/Run issues ONLY!]
Hello there, i'm trying to build my project, but can't go throw linking problem, something i'm doing wrong, but can't understand what exactly, bagging you for help!

My general cmakeLists

```cmake
cmake_minimum_required(VERSION 3.28.3)

project(example			
	LANGUAGES CXX
	VERSION 1.0
)

cmake_policy(SET CMP0077 NEW)
set (CMAKE_CXX_STANDART 11)
add_compile_options(-limm32 -std=c++11  -IImGUI -IImGUI/backends -IImGUI/srrc -Wall -Wunused-parameter -Wformat -Wnarrowing -Wno-sign-conversion -Wno-error-declaration-after-statement)
file (GLOB_RECURSE THIRD_PARTY_SOURCES_FILES ${IMGUI_EXAMPLES_DIR}/libs/* imstb_*.c ../tracy/*.cpp)
set_source_files_properties(${THIRD_PARTY_SOURCES_FILES} PROPERTIES COMPILE_OPTION -w)
add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-class-memaccess>)


if (MSVC)
	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /NODEFAULTLIB:MSVCRT")
endif()

add_subdirectory(SFML)
add_subdirectory(ImGUI)
add_subdirectory(OpenGL)
add_subdirectory(glfw)
add_subdirectory(dependencies)	#[[no this line in tutor]]
add_subdirectory(src)

include_directories(src/include/SFML/include/SFML)
include_directories(/src/include/glad)
#link_directories(/src/include/glad/libglfw3.a)

add_executable (main main.cpp)
#add_executable (main main_test.cpp)

message("test")
#target_link_libraries(main -lopengl32 C:/mingw64/x86_64-w64-mingw32/include/GLFW/lib-mingw-w64/glfw3.dll imm32) 

target_link_libraries(main C:/bones/bones_source/src/imm32.lib imm32)

target_link_libraries(main #C:/bones/bones_source/src/include/imgui/imgui_impl_win32.h imm32
#C:/bones/bones_source/src/include/imgui/imgui_impl_win32.cpp imm32
C:/bones/bones_source/src/include/GLFW/lib-mingw-w64/glfw3.dll imm32
C:/bones/bones_source/src/include/GLFW/lib-mingw-w64/libglfw3.a
C:/bones/bones_source/src/include/GLFW/lib-mingw-w64/libglfw3dll.a)


target_link_libraries(main
#C:/bones/bones_source/lib/libglfw3.a
#C:/bones/bones_source/src/include/GLFW/GLFW.lib imm32
C:/bones/bones_source/src/include/glad/glad.lib 
C:/bones/bones_source/SFML/lib/libFLAC.a
C:/bones/bones_source/SFML/lib/libfreetype.a
C:/bones/bones_source/SFML/lib/libogg.a
C:/bones/bones_source/SFML/lib/libopenal32.a
C:/bones/bones_source/SFML/lib/libsfml-audio-s.a
C:/bones/bones_source/SFML/lib/libsfml-graphics-s.a
C:/bones/bones_source/SFML/lib/libsfml-main.a
C:/bones/bones_source/SFML/lib/libsfml-network-s.a
C:/bones/bones_source/SFML/lib/libsfml-system-s.a
C:/bones/bones_source/SFML/lib/libsfml-window-s.a
C:/bones/bones_source/SFML/lib/libvorbis.a
C:/bones/bones_source/SFML/lib/libvorbisenc.a
C:/bones/bones_source/SFML/lib/libvorbisfile.a
C:/bones/bones_source/ImGUI/imgui.h 
C:/bones/bones_source/ImGUI/imgui.cpp 
C:/bones/bones_source/ImGUI/imgui_demo.cpp
C:/bones/bones_source/ImGUI/imgui_draw.cpp
C:/bones/bones_source/ImGUI/imgui_widgets.cpp
C:/bones/bones_source/src/include/imgui/imgui_impl_glfw.h  
C:/bones/bones_source/src/include/imgui/imgui_impl_glfw.cpp
C:/bones/bones_source/src/include/imgui/imgui_impl_opengl3.h 
C:/bones/bones_source/src/include/imgui/imgui_impl_opengl3.cpp
C:/bones/bones_source/src/include/imgui/imgui_impl_opengl3_loader.h
C:/bones/bones_source/src/include/imgui/imgui_internal.h 
C:/bones/bones_source/src/include/imgui/imgui-SFML.h 
C:/bones/bones_source/src/include/imgui/imgui-SFML.cpp
C:/bones/bones_source/src/include/imgui/imgui-SFML_export.h 
C:/bones/bones_source/src/include/imgui/imstb_rectpack.h 
C:/bones/bones_source/src/include/imgui/imstb_textedit.h 
C:/bones/bones_source/src/include/imgui/imstb_truetype.h 
C:/bones/bones_source/src/include/stb_image.h
)
'''
I'm using the mingw32-make and this is the little part of my erros:
```
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1656b): undefined reference to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1657a): undefined reference to `ImGui::TableSetupScrollFreeze(int, int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1657f): undefined reference to `ImGui::TableHeadersRow()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x16584): undefined reference to `ImGui::TableGetSortSpecs()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x16643): undefined reference to `ImGui::TableNextRow(int, float)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x16648): undefined reference to `ImGui::TableNextColumn()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x16667): undefined reference to `ImGui::TableNextColumn()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x16684): undefined reference to `ImGui::TableNextColumn()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x16698): undefined reference to `ImGui::TableNextColumn()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x166e7): undefined reference to `ImGui::EndTable()'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1740d): undefined reference to `ImGui::BeginTable(char const*, int, int, ImVec2 const&, float)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x17438): undefined reference to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1745c): undefined reference to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x17482): undefined reference to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x174a6): undefined reference to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x174e4): undefined reference to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)'
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1750a): more undefined references to `ImGui::TableSetupColumn(char const*, int, float, unsigned int)' follow
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1751d): undefined reference to `ImGui::TableSetupScrollFreeze(int, int)'
```

**Comment by ocornut** [comment]
You are not specifying what your linking error is.

**Reply by lazzybunny** [reply]
Sorry, miscliked and announced it earlier then i wanted) edit now :)

**Reply by ocornut** [reply]
OK. So what does this error says? Do you seen some sort of pattern in the missing functions? Have you tried searching for those function names in the imgui folders (using "find in files" of your search editor). Can you see what's missing?

**Reply by lazzybunny** [reply]
1.As i understood correctly it says that they missed the cpp files, the compiler don't know what to do with the functions that they called.
2. I've tried to search for the functions, and what weird i can't see this function from the cpp files from witch they called (by the compiler), it feels like i missing some files or something
3. Btw, i was trying to add imgui header files in mingw32 folders and folders that represents in errors, but it doesn't change the situation)) And i was trying to use different PC's, change the compiler three times, rebuild everything few times from scratch, hardcored the inputes in the files, and plays with cmake files a lot, but it doesn't help...
4. Now i feel that i OR miss some imgui files OR linked it in a wrong way

**Reply by ocornut** [reply]
> Have you tried searching for those function names in the imgui folders (using "find in files" of your search editor)

If you search for those function names in imgui codebase you will find they all are in a file called imgui_tables.cpp which you didn’t compile/link into your project.

**Reply by ocornut** [reply]
> Have you tried searching for those function names in the imgui folders (using "find in files" of your search editor)

The first error says:

"C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1656b): undefined reference to ImGui::TableSetupColumn(char const*, int, float, unsigned int)' C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\Users\user\AppData\Local\Temp\ccnPetga.o:imgui_demo.cpp:(.text+0x1657a): undefined reference to ImGui::TableSetupScrollFreeze(int, int)'"

If i search for `TableSetupColumn` in imgui folders:

https://github.com/search?q=repo%3Aocornut%2Fimgui%20TableSetupColumn&type=code

The result is:

![image](https://github.com/user-attachments/assets/2b7e4bcf-6645-46a2-a1db-1f2118dba63a)

**Reply by lazzybunny** [reply]
Thanks a lot, it move my progress a little)
But there is one more problem

And i've got this problem:

[  1%] Building CXX object CMakeFiles/main.dir/main.cpp.obj
[  2%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui_tables.cpp.obj
[  2%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui_demo.cpp.obj
[  3%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui_draw.cpp.obj
[  4%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui_widgets.cpp.obj
[  4%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui_impl_glfw.cpp.obj
[  5%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui_impl_opengl3.cpp.obj
[  6%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui-SFML.cpp.obj
[  6%] Building CXX object CMakeFiles/main.dir/src/include/imgui/imgui.cpp.obj
C:\bones\bones_source\src\include\imgui\imgui.cpp: In constructor 'ImGuiStyle::ImGuiStyle()':
C:\bones\bones_source\src\include\imgui\imgui.cpp:1280:5: error: 'TabBarOverlineSize' was not declared in this scope
 1280 |     TabBarOverlineSize          = 2.0f;             // Thickness of tab-bar overline, which highlights the selected tab-bar.
      |     ^~~~~~~~~~~~~~~~~~
C:\bones\bones_source\src\include\imgui\imgui.cpp:1282:5: error: 'TableAngledHeadersTextAlign' was not declared in this scope; did you mean 'TableAngledHeadersAngle'?
 1282 |     TableAngledHeadersTextAlign = ImVec2(0.5f,0.0f);// Alignment of angled headers within the cell
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     TableAngledHeadersAngle
C:\bones\bones_source\src\include\imgui\imgui.cpp: In member function 'void ImGuiStyle::ScaleAllSizes(float)':
C:\bones\bones_source\src\include\imgui\imgui.cpp:1333:5: error: 'TabBarOverlineSize' was not declared in this scope
 1333 |     TabBarOverlineSize = ImTrunc(TabBarOverlineSize * scale_factor);
'''output

and it's look a little strange. 'cause it seems like imgui.cpp don't see the imgui_table functions, but it was build right above it...

My CMakeLists looks like this 
```cmake
cmake_minimum_required(VERSION 3.28.3)

project(example      
  LANGUAGES CXX
  VERSION 1.0
)

cmake_policy(SET CMP0077 NEW)
set (CMAKE_CXX_STANDART 11)
add_compile_options(-limm32 -std=c++11  -IImGUI -IImGUI/backends -IImGUI/srrc -Wall -Wunused-parameter -Wformat -Wnarrowing -Wno-sign-conversion -Wno-error-declaration-after-statement)
file (GLOB_RECURSE THIRD_PARTY_SOURCES_FILES ${IMGUI_EXAMPLES_DIR}/libs/* imstb_*.c ../tracy/*.cpp)
set_source_files_properties(${THIRD_PARTY_SOURCES_FILES} PROPERTIES COMPILE_OPTION -w)
add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-class-memaccess>)


if (MSVC)
  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /NODEFAULTLIB:MSVCRT")
endif()

add_subdirectory(SFML)
add_subdirectory(ImGUI)
add_subdirectory(OpenGL)
add_subdirectory(glfw)
add_subdirectory(dependencies)  #[[no this line in tutor]]
add_subdirectory(src)

include_directories(src/include/SFML/include/SFML)
include_directories(/src/include/glad)
#link_directories(/src/include/glad/libglfw3.a)

add_executable (main main.cpp
C:/bones/bones_source/src/include/imgui/imgui_tables.cpp
C:/bones/bones_source/src/include/imgui/imgui_demo.cpp
C:/bones/bones_source/src/include/imgui/imgui_draw.cpp
C:/bones/bones_source/src/include/imgui/imgui_widgets.cpp
C:/bones/bones_source/src/include/imgui/imgui_impl_glfw.cpp
C:/bones/bones_source/src/include/imgui/imgui_impl_opengl3.cpp
C:/bones/bones_source/src/include/imgui/imgui-SFML.cpp
C:/bones/bones_source/src/include/imgui/imgui.cpp
)

target_link_libraries(main
C:/bones/bones_source/src/imm32.lib imm32
C:/bones/bones_source/src/include/GLFW/lib-mingw-w64/glfw3.dll imm32
C:/bones/bones_source/src/include/GLFW/lib-mingw-w64/libglfw3.a
C:/bones/bones_source/src/include/GLFW/lib-mingw-w64/libglfw3dll.a
C:/bones/bones_source/src/include/glad/glad.lib 

C:/bones/bones_source/SFML/lib/libFLAC.a
C:/bones/bones_source/SFML/lib/libfreetype.a
C:/bones/bones_source/SFML/lib/libogg.a
C:/bones/bones_source/SFML/lib/libopenal32.a
C:/bones/bones_source/SFML/lib/libsfml-audio-s.a
C:/bones/bones_source/SFML/lib/libsfml-graphics-s.a
C:/bones/bones_source/SFML/lib/libsfml-main.a
C:/bones/bones_source/SFML/lib/libsfml-network-s.a
C:/bones/bones_source/SFML/lib/libsfml-system-s.a
C:/bones/bones_source/SFML/lib/libsfml-window-s.a
C:/bones/bones_source/SFML/lib/libvorbis.a
C:/bones/bones_source/SFML/lib/libvorbisenc.a
C:/bones/bones_source/SFML/lib/libvorbisfile.a

C:/bones/bones_source/src/include/imgui/imgui.h 
C:/bones/bones_source/src/include/imgui/imgui_impl_glfw.h  
C:/bones/bones_source/src/include/imgui/imgui_impl_opengl3.h 
C:/bones/bones_source/src/include/imgui/imgui_impl_opengl3_loader.h
C:/bones/bones_source/src/include/imgui/imgui_internal.h 

C:/bones/bones_source/src/include/imgui/imgui-SFML.h 
C:/bones/bones_source/src/include/imgui/imgui-SFML_export.h 
C:/bones/bones_source/src/include/imgui/imstb_rectpack.h 
C:/bones/bones_source/src/include/imgui/imstb_textedit.h 
C:/bones/bones_source/src/include/imgui/imstb_truetype.h 


C:/bones/bones_source/src/include/stb_image.h
)
```

**Reply by ocornut** [reply]
This seems like your different source/include files are from different version of dear imgui.

**Reply by ocornut** [reply]
I'm not sure I understand why people feel the urge to copy and move files around and create this sorts of issue for themselves.
It's clear from your paths that you are not using an unmodified imgui/ repository. If you are not familiar with C++ building tools it may be better to use unmodified libraries rather than move things around, in my opinion.

**Reply by lazzybunny** [reply]
I don't know, sorry for that) i think it's just the classical problem of someone who are new to something and just trying to solve the problem and explore the way to programming things by himself, but instead making more issues) 
Thanks a lot for your advice, i changed the version of every part of the project, and i'm not sure, but it seems that "imgui issues" has gone, and it seems like a progress, 'cause the building progress changed from 7% to 70))
